src/cluster_legacy.c

```diff
@@ -6165,61 +6416,128 @@ int clusterCommandSpecial(client *c) {
+                                         CLUSTER_TODO_FSYNC_CONFIG);
+                }
+
+                /* If this node or this node's primary was importing this slot,
+                * assigning the slot to itself also clears the importing status. */
+                if ((n == myself || n == myself->slaveof) &&
+                    server.cluster->importing_slots_from[slot])
+                {
+                    server.cluster->importing_slots_from[slot] = NULL;
+
+                    /* Only primary broadcasts the updates */
+                    if (n == myself) {
+                        /* This slot was manually migrated, set this node configEpoch
+                         * to a new epoch so that the new version can be propagated
+                         * by the cluster.
+                         *
+                         * Note that if this ever results in a collision with another
+                         * node getting the same configEpoch, for example because a
+                         * failover happens at the same time we close the slot, the
+                         * configEpoch collision resolution will fix it assigning
+                         * a different epoch to each node. */
+                        if (clusterBumpConfigEpochWithoutConsensus() == C_OK) {
+                            serverLog(LL_NOTICE,
+                                      "ConfigEpoch updated after importing slot %d",
+                                      slot);
+                        }
+                        /* After importing this slot, let the other nodes know as
+                         * soon as possible. */
+                        clusterBroadcastPong(CLUSTER_BROADCAST_ALL);
+                    }
                 }
-                server.cluster->importing_slots_from[slot] = NULL;
-                /* After importing this slot, let the other nodes know as
-                 * soon as possible. */
-                clusterBroadcastPong(CLUSTER_BROADCAST_ALL);
+
+                /* Don't replicate again if setslot has been replicated */
+                replSetSlot &= !replDone;
+            } else {
+                /* We are a primary and this is the first time we see this "setslot"
+                 * command. Force-replicate the setslot command to all of our replicas
+                 * first and only on success will we handle the command.
+                 * Note that
+                 * 1. All replicas are expected to ack the replication within 1000ms
+                 * 2. The repl offset target is set to the master's current repl offset + 1.
+                 *    There is no concern of partial replication because replicas always
+                 *    ack the repl offset at the command boundary. */
+                blockForPreReplication(c, mstime()+1000, server.master_repl_offset+1, myself->numslaves);
+                replicationRequestAckFromSlaves();
+
+                /* Don't reply to the client yet */
```

Comment:
Would be good to add comment, why we don't reply to the client in this case and when/where does it get sent to complete the command. I guess the client would be hung till then.

Reply:
Yes - the client will be hung. Will do.